ee/katc/structured_clone.go

```diff
@@ -0,0 +1,275 @@
+package katc
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"strconv"
+
+	"golang.org/x/text/encoding/unicode"
+	"golang.org/x/text/transform"
+)
+
+const (
+	tagHeader        uint32 = 0xfff10000
+	tagNull          uint32 = 0xffff0000
+	tagUndefined     uint32 = 0xffff0001
+	tagBoolean       uint32 = 0xffff0002
+	tagInt32         uint32 = 0xffff0003
+	tagString        uint32 = 0xffff0004
+	tagArrayObject   uint32 = 0xffff0007
+	tagObjectObject  uint32 = 0xffff0008
+	tagBooleanObject uint32 = 0xffff000a
+	tagStringObject  uint32 = 0xffff000b
+	tagEndOfKeys     uint32 = 0xffff0013
+)
+
+// structuredCloneDeserialize deserializes a JS object that has been stored by Firefox
+// in IndexedDB sqlite-backed databases.
+// References:
+// * https://stackoverflow.com/a/59923297
+// * https://searchfox.org/mozilla-central/source/js/src/vm/StructuredClone.cpp (see especially JSStructuredCloneReader::read)
+func structuredCloneDeserialize(ctx context.Context, slogger *slog.Logger, row map[string][]byte) (map[string][]byte, error) {
```

Comment:
To keep this PR from expanding too much further, I will write unit tests for this file in a follow-up PR. A decent portion of this file is already exercised via `Test_generate_SqliteBackedIndexedDB`.